lib/uktest: insert .uk_testtab after data section

alexhoppus opened the pull request:
the problem with inserting uk_testtab after .rodata, is that uk_testtab appears on the same page with .rodata, what means .uk_testtab would be write protected in some cases, that may lead to the data abort when executing libuktest macros.

```
[    0.000000] Info: [libukboot] <boot.c @  199> Unikraft constructor table at 0x40126000 - 0x40126030
test: uktest_myself_testsuite->uktest_test_sanity
    :	expected `((void *) 0)` to be 0 and was 0 ....................................... [ PASSED ]
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  192> EL1 sync trap caught
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  158> 	 SP       : 0x0000000047fbccc0
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  159> 	 ESR_EL1  : 0x000000009600004f
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  160> 	 ELR_EL1  : 0x000000004011cef8
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  161> 	 LR (x30) : 0x000000004011ce5c
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  162> 	 PSTATE   : 0x00000000800003c5
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  163> 	 FAR_EL1  : 0x000000004012c7ac
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  166> 	 x00 ~ x03: 0x0000000000000001 0x000000004012c7ac 0x0000000000000000 0x000000000000002a
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  166> 	 x04 ~ x07: 0x000000000000ffff 0x0000000047fbc997 0x0000000000000067 0x000000000000000a
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  166> 	 x08 ~ x11: 0x0000000000000073 0x0000000000000000 0x000000004011115c 0x0000000000000000
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  166> 	 x12 ~ x15: 0x0000000047fbcde0 0x0000000000000020 0x00000000ffffffe8 0x0000000000000020
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  166> 	 x16 ~ x19: 0x0000000000000000 0x0000000000000000 0x0000000047fbc8a8 0x000000004012c750
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  166> 	 x20 ~ x23: 0x000000000000002a 0x000000004012c7e4 0x000000004012c430 0x000000004012c448
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  166> 	 x24 ~ x27: 0x0000000040129b58 0x000000004012c7df 0x000000004012b0d8 0x000000004012e000
[    0.000000] CRIT: [libkvmplat] <traps_arm64.c @  170> 	 x28 ~ x29: 0x0000000000000000 0x0000000047fbcde0
```
the problem arises on arm64 if paging & w_xor_x options are enabled.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Description of changes

the ukttestab section moved in linker script to be included after data section.

Commit:
lib/uktest: insert .uk_testtab after data section

the problem with inserting after .rodata, is that uk_testtab appears on
the same page with .rodata, what means .uk_testtab would be write
protected, that leads to the data abort when executing libuktest macros

Signed-off-by: Aleksandr Iashchenko <aia@opensynergy.com>

alexhoppus commented:
The alternative (and maybe right) way is to force __PAGE_SIZE alignment for this section, but that increases the size of the image.

razvanvirtan reviewed (approved):
Hi Alex, thank you for this!
The bug is reproducible. I tested the PR in combination with https://github.com/unikraft/unikraft/pull/439 and it works as expected.
If @marcrittinghaus doesn't prefer the `__PAGE_SIZE` alignment alternative you mentioned, I think we can merge this.

Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

marcrittinghaus reviewed (approved):
Makes sense. Thank you very much for the bug fix! 😃

razvand commented:
Hi, @alexhoppus . Once I apply this PR, tests won't run on x86. To try this, use the [`do-libsodium` test script](https://github.com/unikraft-upb/scripts/tree/main/musl#libsodium):

```
$ ./do-libsodium setup
$ ./do-libsodium build
$ ./do-libsodium run
```

Tests run.

But, if you set the [`need_x86_tests` variable](https://github.com/unikraft-upb/scripts/blob/main/musl/include/common_functions#L3) to `0`, and then run:

```
$ ./do-libsodium remove
$ ./do-libsodium build
$ ./do-libsodium run
```

Then tests won't run.

It's pretty weird this happens, since this is a very small commit. @kubanrob can confirm it.

@marcrittinghaus , any idea why this would happen? I recall you worked on `uktest`.

kubanrob commented:
I investigated a bit further now, for some reason the `cases_end` field of the test cases is empty.
```
uk_pr_crit("in testsuite %p before case %p to %p\n", suite,_UK_TESTCASE_START(suite),(suite)->cases_end)

Output:
[...]
[    0.178248] dbg:  [libukboot] <boot.c @  226> Call init function: 0x1175b0() at 0x11b028...
[    0.182586] CRIT: [libuktest] <test.c @  151> in testsuite 0x121100
[    0.186839] CRIT: [libuktest] <test.c @  174> in testsuite 0x121100 before case 0x12111c to 0
[...]

$ objdump -t build/app-helloworld_kvm-x86_64.dbg | grep -E "uk_libc_testsuite" | sort
[...]
000000000011b028 l     O .uk_inittab	0000000000000008 __uk_inittab69_testsuite_run_uk_libc_testsuite
0000000000121100 l     O .uk_testtab	000000000000001c testsuite_uk_libc_testsuite
000000000012111c l     O .uk_testtab	0000000000000028 _uk_testtab_uk_libc_testsuite_string_memcpy_tests
0000000000121144 l     O .uk_testtab	0000000000000028 _uk_testtab_uk_libc_testsuite_string_strchr_tests
000000000012116c l     O .uk_testtab	0000000000000028 _uk_testtab_uk_libc_testsuite_string_strcspn_tests
0000000000121194 l     O .uk_testtab	0000000000000028 _uk_testtab_uk_libc_testsuite_string_strstr_tests
00000000001211bc l       .uk_testtab	0000000000000000 testsuite_run_uk_libc_testsuite__end
[...]
```

razvand commented:
> I investigated a bit further now, for some reason the `cases_end` field of the test cases is empty.

Nice find, @kubanrob . There may be some default zero initialization? That should be the case with the `.bss` section, not the `.data` one, however.

kubanrob commented:
@razvand You're probably right, the name is zero also ... and .bss follows the .data section.
```
uk_pr_crit("in testsuite %s (%p) before case %p to %p\n", suite->name,suite,_UK_TESTCASE_START(suite),(suite)->cases_end);

$ objdump -t build/app-helloworld_kvm-x86_64.dbg | grep -E "uk_libc_testsuite|bss|data" | sort
0000000000121100 g       .data	0000000000000000 _edata
0000000000121100 l     O .uk_testtab	000000000000001c testsuite_uk_libc_testsuite
000000000012111c l     O .uk_testtab	0000000000000028 _uk_testtab_uk_libc_testsuite_string_memcpy_tests
0000000000121144 l     O .uk_testtab	0000000000000028 _uk_testtab_uk_libc_testsuite_string_strchr_tests
000000000012116c l     O .uk_testtab	0000000000000028 _uk_testtab_uk_libc_testsuite_string_strcspn_tests
0000000000121194 l     O .uk_testtab	0000000000000028 _uk_testtab_uk_libc_testsuite_string_strstr_tests
00000000001211bc l       .uk_testtab	0000000000000000 testsuite_run_uk_libc_testsuite__end
00000000001211bc l       .uk_testtab	0000000000000000 _uk_testsuite_uk_libc_testsuite_string_memcpy_tests__asserts_begin
00000000001211c8 l       .uk_testtab	0000000000000000 _uk_testsuite_uk_libc_testsuite_string_strchr_tests__asserts_begin
0000000000121280 l       .uk_testtab	0000000000000000 _uk_testsuite_uk_libc_testsuite_string_strcspn_tests__asserts_begin
00000000001212a4 l       .uk_testtab	0000000000000000 _uk_testsuite_uk_libc_testsuite_string_strstr_tests__asserts_begin
0000000000122000 g       .bss	0000000000000000 __bss_start

[    0.185144] CRIT: [libuktest] <test.c @  174> in testsuite (null) (0x121100) before case 0x12111c to 0
```

kubanrob commented:
plat/kvm/x86/multiboot.S
```
 48 multiboot_header:
 49         .long MULTIBOOT_HEADER_MAGIC
 50         .long MULTIBOOT_FLAGS
 51         .long -(MULTIBOOT_HEADER_MAGIC+MULTIBOOT_FLAGS) /* checksum */
 52         .long multiboot_header                          /* header addr */
 53         .long 0x100000                                  /* load addr */
 54         .long _edata                                    /* load end addr */
 55         .long _end                                      /* bss end addr */
 56         .long multiboot_start32                         /* entry addr */
```
The testcases are now between `_edata` and `_end` I assume the bootloader zeros it or it is not loaded by at all.

kubanrob commented on lib/uktest/tests.ld:
```diff
@@ -9,4 +9,4 @@ SECTIONS
 	}
 }
-INSERT AFTER .rodata;
+INSERT AFTER .data;
```
I tested it with `.uk_inittab` (test run in x86_64 and aarch64), but probably any of the sections between `_etext` and `_rodata` is fine.
```suggestion
INSERT AFTER .uk_inittab;
```

alexhoppus replied:
Sorry, I don't understand how this fixes initial problem. From my point of view, we could go with two solutions 1) roll back to PAGE_SIZE alignment solution 2) change multiboot_header : load_end_addr should point to _bss_start roughly

kubanrob replied:
After a short discussion: I agree with @alexhoppus that after `.data` is a good place for this section and the multiboot header should get the start of the `.bss` section as `load_end_addr` (option 2). The `.uk_inittab` tables are also read-only, it is only that they are currently not protected as they have no memory region in `plat/kvm/memory.c`.

Maybe outside the scope of this PR: I think in the future it would be a good idea to mark specific locations in the linkerscript where and how extra read-writable and read-only data section can safely placed by linkerscripts includes.

marcrittinghaus replied:
Sorry, for not jumping into this discussion earlier. I did test the change and it worked for me. But yeah, different configurations may lead to different results. This is really not optimal. I agree @kubanrob.

Actually, for changing the multiboot header I already have a fix internally as I had a similar issue some time ago (though not with the uktest section). I can provide the patch as PR, if you like. But go ahead if you have it ready.

marcrittinghaus commented:
@kubanrob, @alexhoppus Shall I provide the PR for the multiboot fix or will you do that?

I would like to have that PR also ready before I add my `Approved-by` again.

Commit:
plat/kvm/x86: change multiboot header load end addr

in order to load sections after .data section redefine load_end_addr
in multiboot header to _bss_start

Signed-off-by: Aleksandr Iashchenko <aia@opensynergy.com>

alexhoppus commented:
@marcrittinghaus sorry for delay, it seems this load_end_addr -> __bss_start works for me. I have not checked with ./do-libsodium script, however I think I have reproduced the same just with x86 config + enabled uktests. So, please, if this is okay for you, you can use my changes. If you don't like it or for some reason it doesn't work, please, we can go with your PR for multiboot. Thank you.

marcrittinghaus commented:
> @marcrittinghaus sorry for delay, it seems this load_end_addr -> __bss_start works for me. I have not checked with ./do-libsodium script, however I think I have reproduced the same just with x86 config + enabled uktests. So, please, if this is okay for you, you can use my changes. If you don't like it or for some reason it doesn't work, please, we can go with your PR for multiboot. Thank you.

Your updated version is exactly what I did so thank you for adding this to the PR 😃 @razvand Could you retry with your script if it now works for you?

kubanrob reviewed (approved):
Now works me on x86_64 as well as aarch64. Thank you @alexhoppus!

Reviewed-by: Robert Kuban <robert.kuban@opensynergy.com>

marcrittinghaus reviewed (approved):
Although "...libuktest macroses" sounds like uktest has some serious illness 😄, I think this is good to go. Thanks for the PR @alexhoppus and for testing @kubanrob @razvand @razvanvirtan 🥇

Approved-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>
